### PR TITLE
BTC singing error codes exposed

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

/Updated/
Bitcoin SigningOutput (and also TransactionPlan) had a free-text error field.  Changed this into a numerical code (enum), with possible values listed as an enum in the proto file.
This is a breaking change.

For hints on updating mobile apps, see updated tests:
https://github.com/trustwallet/wallet-core/pull/1317/files#diff-d06f41853ba3b8aa2c036da5b66b5bc993fcc5134c03117bcfc4762d7636fc9d
https://github.com/trustwallet/wallet-core/pull/1317/files#diff-3a651a2e25b415b7c0c5381f21b4aff94e1d75c6e93599773d65e928d3766e78

## Testing instructions

Unit tests

## Types of changes

* Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Decred/Signer.h [hunk 1]
```diff
@@ -65,15 +66,15 @@ class Signer {
     /// Signs the transaction.
     ///
     /// \returns the signed transaction or an error.
-    Result<Transaction> sign();
+    Result<Transaction, TW::Bitcoin::Error> sign();
```
Comment (hewigovens) [inline src/Decred/Signer.h:69]:
```suggestion
    Result<Transaction, Bitcoin::Error> sign();
```

File: src/Decred/Signer.h [hunk 2]
```diff
@@ -65,15 +66,15 @@ class Signer {
...
 
     /// Signs a particular input.
     ///
     /// \returns the signed transaction script.
-    Result<Bitcoin::Script> sign(Bitcoin::Script script, size_t index);
+    Result<Bitcoin::Script, TW::Bitcoin::Error> sign(Bitcoin::Script script, size_t index);
```
Comment (hewigovens) [inline src/Decred/Signer.h:74]:
```suggestion
    Result<Bitcoin::Script, Bitcoin::Error> sign(Bitcoin::Script script, size_t index);
```

File: src/Decred/Signer.h [hunk 3]
```diff
@@ -65,15 +66,15 @@ class Signer {
...
 
   private:
-    Result<std::vector<Data>> signStep(Bitcoin::Script script, size_t index);
+    Result<std::vector<Data>, TW::Bitcoin::Error> signStep(Bitcoin::Script script, size_t index);
```
Comment (hewigovens) [inline src/Decred/Signer.h:77]:
```suggestion
    Result<std::vector<Data>, Bitcoin::Error> signStep(Bitcoin::Script script, size_t index);
```

File: src/proto/FIO.proto [hunk 4]
```diff
@@ -150,11 +150,25 @@ message SigningInput {
     Action action = 5;
 }
 
+enum ErrorCode {
```
Comment (hewigovens) [inline src/proto/FIO.proto:153]: can FIO reuse EOS `ErrorCode`?

**Review by hewigovens — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]
Overall looks good, one suggestion

- Explore about Universal format for error text / status

**Comment by catenocrypt** [comment]
Back to Draft for some simplifications

File: src/Decred/Signer.h (hunk 1 above)
Comment (catenocrypt) [reply src/Decred/Signer.h]: ok, done

File: src/proto/FIO.proto (hunk 4 above)
Comment (catenocrypt) [reply src/proto/FIO.proto]: now there is one common def, each chain references it

**Comment by catenocrypt** [comment]
Reworked, simplified:
- there is only an error code now (no free text field)
- there is a common error code enum for all chains

**Review by hewigovens — APPROVED** [review]

**Comment by catenocrypt** [comment]
Note on diff coverage result: obviously there are many error-returning lines touched, and many of them are uncovered.  But overall coverage is not decreased.

**Review by ArtemGoryunov — APPROVED** [review]

**Pull request merged**